### PR TITLE
Added Comments for CppInterOp's Doxygen based documentation

### PR DESCRIPTION
**Pull request by QuillPusher**
I have added function definitions as Doxygen-style comments, based on the CppInterop information session between Baidyanath and Krishna.

Please review if my understanding was correct in these definitions (since I'm new to the domain).

Requested Reviewers: @vgvassilev  @sudo-panda

### CODE REVIEW

File: include/clang/Interpreter/InterOp.h [hunk 1]
```diff
@@ -1,5 +1,5 @@
 //--------------------------------------------------------------------*- C++ -*-
-// CLING - the C++ LLVM-based InterpreterG :)
+// CLING - the C++ LLVM-based Interpreter :)
```
Comment (vgvassilev) [inline include/clang/Interpreter/InterOp.h:2]: We should change the file preamble separately. I can do that later outside of this PR.

File: include/clang/Interpreter/InterOp.h [hunk 2]
```diff
@@ -59,7 +59,11 @@ namespace InterOp {
       : m_Kind(K), m_GenericCall(C), m_FD(FD) {}
     JitCall(Kind K, DestructorCall C, TCppConstFunction_t Dtor)
       : m_Kind(K), m_DestructorCall(C), m_FD(Dtor) {}
+	  
+	/// Check if the passed arguments are valid for the given function.
```
Comment (vgvassilev) [inline include/clang/Interpreter/InterOp.h:63]: Seems the indentation is a bit off. Could you run `git-clang-format` to the PR?

File: include/clang/Interpreter/InterOp.h [hunk 3]
```diff
@@ -78,10 +82,10 @@ namespace InterOp {
     /// Makes a call to a generic function or method.
     ///\param[in] result - the location where the return result will be placed.
     ///\param[in] args - a pointer to a argument list and argument size.
-    ///\param[in] self - the this pointer of the object.
```
Comment (vgvassilev) [inline include/clang/Interpreter/InterOp.h:81]: The `this pointer` is a term we should keep it.

File: include/clang/Interpreter/InterOp.h [hunk 4]
```diff
@@ -121,163 +125,267 @@ namespace InterOp {
   ///\returns true if the debugging printouts on stderr are enabled.
   bool IsDebugOutputEnabled();
 
+  /// Identifies if the class in its scope is of aggregate type, 
+  /// or not (for example, simple C-Style struct classes that do not
+  /// use OOP).
```
Comment (vgvassilev) [inline include/clang/Interpreter/InterOp.h:130]:
```suggestion
  /// Identifies if the given class represents an aggregate type).
  ///\returns true if \c scope is an array or a C++ tag (as per C++ [dcl.init.aggr])
```

File: include/clang/Interpreter/InterOp.h [hunk 5]
```diff
@@ -121,163 +125,267 @@ namespace InterOp {
...
+  /// Identifies if the scope is a namespace or not.
   bool IsNamespace(TCppScope_t scope);
 
+  /// Identifies if the scope is a class or not.
   bool IsClass(TCppScope_t scope);
+  
   // See TClingClassInfo::IsLoaded
+  /// Identifies if the class definition is present, or not 
+  /// (e.g., in case a header is included to declare the class,
+  /// but the class is not defined).
```
Comment (vgvassilev) [inline include/clang/Interpreter/InterOp.h:143]:
```suggestion
  /// Identifies if the class definition is present, or not. Performs a
  /// template instantiation if necessary. 
```

File: include/clang/Interpreter/InterOp.h [hunk 6]
```diff
@@ -121,163 +125,267 @@ namespace InterOp {
...
   size_t SizeOf(TCppScope_t scope);
+  
+  /// Identifies if it is a "built-in" or a "complex" type.
   bool IsBuiltin(TCppType_t type);
 
+  /// Identifies if it is a templated class.
   bool IsTemplate(TCppScope_t handle);
 
+  /// Identifies if it is a templated class and it also has
+  ///  the template arguments.
```
Comment (vgvassilev) [inline include/clang/Interpreter/InterOp.h:155]:
```suggestion
  /// Identifies if it is a class template specialization class.
```

**Review by vgvassilev — COMMENTED** [review]
Here are some comments from my partial review. Perhaps we could use `Check` instead of `Identify`.

@sudo-panda, can you go through the documentation and check if it is aligned with the implementation?

File: include/clang/Interpreter/InterOp.h (hunk 3 above)
Comment (QuillPusher) [reply include/clang/Interpreter/InterOp.h:81]: changed to
`param[in] self - the 'this pointer' of the object.`

File: include/clang/Interpreter/InterOp.h (hunk 1 above)
Comment (QuillPusher) [reply include/clang/Interpreter/InterOp.h:2]: change reverted

File: include/clang/Interpreter/InterOp.h (hunk 2 above)
Comment (QuillPusher) [reply include/clang/Interpreter/InterOp.h]: done in 8th new commit

File: include/clang/Interpreter/InterOp.h [hunk 7]
```diff
@@ -121,179 +126,282 @@ namespace InterOp {
...
   TCppType_t GetEnumConstantType(TCppScope_t scope);
 
+  /// Gets the index value (0,1,2, etcetera) of the enum constant
+  /// that was passed into this function.
   TCppIndex_t GetEnumConstantValue(TCppScope_t scope);
 
+  /// Gets the size of the "type" that is passed in to this function.
   size_t GetSizeOfType(TCppType_t type);
 
+  /// Checks  if the passed value is a variable.
```
Comment (vgvassilev) [inline include/clang/Interpreter/InterOp.h:196]:
```suggestion
  /// Checks if the passed value is a variable.
```

File: include/clang/Interpreter/InterOp.h [hunk 8]
```diff
@@ -121,179 +126,282 @@ namespace InterOp {
...
   bool IsConstVariable(TCppScope_t var);
 
+  /// Checks if the provided parameter is a Record (struct).
   bool IsRecordType(TCppType_t type);
 
+  /// Checks if the provided parameter is a Plain Old Data Type (POD).
   bool IsPODType(TCppType_t type);
 
+  /// Gets the pure, Underlying Type (as opposed to the Using Type).
   TCppType_t  GetUnderlyingType(TCppType_t type);
```
Comment (vgvassilev) [inline include/clang/Interpreter/InterOp.h:370]:
```suggestion
  TCppType_t GetUnderlyingType(TCppType_t type);
```

File: lib/Interpreter/InterOp.cpp [hunk 9]
```diff
@@ -822,6 +825,8 @@ namespace InterOp {
     return "<unknown>";
   }
 
+  // InterOp's internal functions that are not needed outside the library are
```
Comment (vgvassilev) [inline lib/Interpreter/InterOp.cpp:828]:
```suggestion
  // Internal functions that are not needed outside the library are
```

File: lib/Interpreter/InterOp.cpp [hunk 10]
```diff
@@ -1173,6 +1181,9 @@ namespace InterOp {
       return QT.getCanonicalType().getAsOpaquePtr();
   }
 
+  // InterOp's internal functions that are not needed outside the library are
```
Comment (vgvassilev) [inline lib/Interpreter/InterOp.cpp:1184]:
```suggestion
  // Internal functions that are not needed outside the library are
```

File: lib/Interpreter/InterOp.cpp [hunk 11]
```diff
@@ -1256,13 +1267,17 @@ namespace InterOp {
     return C.getTypeDeclType(cast<TypeDecl>(D)).getAsOpaquePtr();
   }
 
+  // InterOp's internal functions that are not needed outside the library are
```
Comment (vgvassilev) [inline lib/Interpreter/InterOp.cpp:1270]:
```suggestion
  // Internal functions that are not needed outside the library are
```

**Comment by QuillPusher** [comment]
Squashed last 10 commits and summarized all comments into the last force-push commit, no new changes to the files

**Review by vgvassilev — APPROVED** [review]
LGTM!

**Pull request merged**